core/src/main/scala-2.13+/cats/data/NonEmptyLazyList.scala

```diff
@@ -121,6 +121,8 @@ class NonEmptyLazyListOps[A](private val value: NonEmptyLazyList[A])
   /**
    * Alias for [[prepend]].
    */
+  // TODO: `a` should be by-name and this method should not be listed as an
+  //       alias for `prepend`, but it's too late to change that in this version
   final def #::[AA >: A](a: AA): NonEmptyLazyList[AA] =
     prepend(a)
```

Comment:
in case clarification is needed, the API in `LazyList` which this is attempting to mirror is by-name (in fact, doubly by-namee), so this should be too

Reply:
Why is it too late to change that? Binary-compatibility? We can change it.

Reply:
if that's the case, then there are further API changes that are worth discussing as well. I'll try to gather up some links and thoughts when I get back to my computer

Reply:
so, `#::` for `LazyList` is actually defined via implicit conversion that makes the tail part lazy as well (see scala/scala#8985 for the nuances of doing this properly). ideally, `#::` for `NonEmptyLazyList` would be implemented the same way, but I don't know just how much is changeable (I still don't understand why bincompat isn't important here).

on top of this, `fromLazyListPrepend` and `fromLazyListAppend` are rather poorly designed. both parameters to both methods ought to be by-name to give users the ability to make fully lazy `NonEmptyLazyList`s. additionally, as mentioned in another comment, `apply` is very un-lazy. `appendLazyList` and `prependLazyList` probably also ought to both have by-name parameters for better laziness (similar to `LazyList#lazyAppendedAll`). the same with `concatNell`, `prependNell`, and their various symbolic aliases

Reply:
> I still don't understand why bincompat isn't important here

Bincompat is always important :) but in many cases it is possible to evolve APIs without breaking bincompat.

Why don't you go ahead and make all the changes you think should happen? Then we can go through them case-by-case and fixup bincompat issues.

Reply:
with a blessing like that, how can I refuse?

Reply:
@armanbilge done. if significant parts of this go places, it will be very helpful to be able to use the framework I wrote for testing laziness in scala/scala. does this repo's license/typelevel's policy permit borrowing code and altering it like that?

Reply:
Cats License is MIT, and some of our own codebase is adapted from scala/scala.

Reply:
@NthPortal I just pushed 9c4a5ec143a958d15147cfcbc9458b10652fa236 which I think restores bincompat, at least on 2.13. 2.12 will probably need more massaging but we'll see. Anyway, that commit should demonstrate the general techniques, feel free to ask questions :)

I'm not 100% sure if I fixed everything because I was getting a lot of spurious MiMa errors. I think your branch is out of date and needs a merge or rebase, would you mind doing that? Thanks!

Reply:
Ha, looks like CI ate that up very nicely. Well, there you go :)

Reply:
> I think your branch is out of date

understatement, I started this ages ago, but didn't didn't get around to writing tests until this weekend

Reply:
there is one thing that still needs fixing—`#::`. will the empty param list trick work with that as well?